internal/replication/blockreplicator_test.go

```diff
@@ -466,3 +467,169 @@ func TestBlockReplicator_ReConfig(t *testing.T) {
+		t.Logf("Snap list: %v", snapList)
+	})
+
+	t.Run("restart from a snapshot", func(t *testing.T) {
+		lg := testLogger(t, "debug")
+		testDir, err := ioutil.TempDir("", "replication-test")
+		require.NoError(t, err)
+		defer os.RemoveAll(testDir)
+
+		block := &types.Block{
+			Header: &types.BlockHeader{
+				BaseHeader: &types.BlockHeaderBase{
+					Number:                1,
+					LastCommittedBlockNum: 1,
+				},
+			},
+			Payload: &types.Block_DataTxEnvelopes{},
+		}
+		blockData, err := proto.Marshal(block)
+		require.NoError(t, err)
+		blockLen := uint64(len(blockData))
+
+		clusterConfig := proto.Clone(clusterConfig1node).(*types.ClusterConfig)
+		clusterConfig.ConsensusConfig.RaftConfig.SnapshotIntervalSize = 3*blockLen + 1 // take a snapshot every 4 data blocks
+		env, err := newNodeEnv(1, testDir, lg, clusterConfig)
+		require.NoError(t, err)
+		require.NotNil(t, env)
+
+		err = env.Start()
+		require.NoError(t, err)
+
+		// wait for the node to become leader
+		isLeaderCond := func() bool {
+			return env.blockReplicator.IsLeader() == nil
+		}
+		assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+		numBlocks := uint64(20)
+		for i := uint64(0); i < numBlocks; i++ {
+			b := proto.Clone(block).(*types.Block)
+			b.Header.BaseHeader.Number = 2 + i
+			err := env.blockReplicator.Submit(b)
+			require.NoError(t, err)
+
+		}
+		assert.Eventually(t, func() bool {
+			h, err := env.ledger.Height()
+			if err == nil && h == numBlocks+1 {
+				return true
+			}
+			return false
+		}, 30*time.Second, 100*time.Millisecond)
+
+		err = env.Close()
+		require.NoError(t, err)
+
+		snapList := replication.ListSnapshots(env.conf.Logger, env.conf.LocalConf.Replication.SnapDir)
+		require.Equal(t, 4, len(snapList))
+		t.Logf("Snap list after close: %v", snapList)
+
+		//recreate - snapshot and no following entries
+		err = env.Restart()
+		require.NoError(t, err)
+
+		isLeaderCond = func() bool {
+			return env.blockReplicator.IsLeader() == nil
+		}
+		assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+		// add 6 blocks, to create another snapshot and 2 more entries
+		numBlocksExtra := uint64(6)
+		for i := numBlocks; i < numBlocks+numBlocksExtra; i++ {
+			b := proto.Clone(block).(*types.Block)
+			b.Header.BaseHeader.Number = 2 + i
+			err := env.blockReplicator.Submit(b)
+			require.NoError(t, err)
+
+		}
+		assert.Eventually(t, func() bool {
+			h, err := env.ledger.Height()
+			if err == nil && h == numBlocks+numBlocksExtra+1 {
+				return true
+			}
+			return false
+		}, 30*time.Second, 100*time.Millisecond)
+
+		err = env.Close()
+		require.NoError(t, err)
+
+		snapList2 := replication.ListSnapshots(env.conf.Logger, env.conf.LocalConf.Replication.SnapDir)
+		require.Equal(t, 4, len(snapList2))
+		t.Logf("Snap list after 2nd close: %v", snapList2)
+		require.Equal(t, snapList[3], snapList2[2])
+
+		//recreate - snapshot and two following entries
+		err = env.Restart()
+		require.NoError(t, err)
```

Comment:
I think I missing something, but don't we need to check is `env` restarted from correct snapshot?

Reply:
Good question. In general, the node recovers correctly if it can continue to function: i.e. elect a leader and agree on block order. If block order after recovery is messed up (e.g. trying to `Append` the same block twice or `Append` with a gap), the mock ledger will panic and the test will fail.

In order to ensure the above, I added a few block submissions after this part (second recovery), and check that they get to the ledger ok. 

I also added a few comments on each of the test cases to make it easier to read.